crates/language/src/target_language.rs

```diff
@@ -173,6 +185,88 @@ impl PatternLanguage {
         Self::get_language_with_parser(&mut parser, src)
     }
 
+    fn get_aliases(&self) -> &'static [&'static str] {
+        for (variant, aliases) in LANGUAGE_ALIASES {
+            if variant == self {
+                return aliases;
+            }
+        }
+        &[]
+    }
+
+    fn map_value(s: String) -> Self {
+        if let Some(lang) = Self::from_string(&s, None) {
+            return lang;
+        }
+        if let Some(lang) = Self::from_extension(&s) {
+            return lang;
+        }
+        let s = s.to_lowercase();
+        for (variant, aliases) in LANGUAGE_ALIASES {
+            for alias in *aliases {
+                if *alias == s {
+                    return *variant;
+                }
+            }
+        }
+        Self::Universal
+    }
+
+    fn possible_values() -> Vec<&'static str> {
+        let mut unique_variants = std::collections::HashSet::new();
+        let mut ordered_variants = Vec::new();
```

Comment:
Two collections is unnecessary, let's just build the hashset then convert to Vec at the end.

Reply:
I had used a `HashSet` alone originally like this:

```rust
    fn possible_values() -> Vec<&'static str> {
        let mut unique_variants = std::collections::HashSet::new();
        for lang in PatternLanguage::enumerate() {
            let lang_name = lang.to_str();
            unique_variants.insert(lang_name);
            let aliases = lang.get_aliases();
            for alias in aliases {
                unique_variants.insert(alias);
            }
            for ext in lang.get_file_extensions() {
                unique_variants.insert(ext);
            }
        }
        unique_variants.into_iter().collect()
    }
```

which works but the insertion order is lost and so the order of print out of possible values that is displayed looks weird IMO because it is shuffled. I personally thought that it looked better when the name / aliases / extensions appeared grouped/together like this screenshot I shared [above](https://github.com/getgrit/gritql/pull/544#issuecomment-2425149258):
```
echo 'console.log(hello)' | cargo run -- apply '`hello` => `goodbye`' --stdin --lang=javascrip
```
![image](https://github.com/user-attachments/assets/bbb2d9fd-aea1-4d57-9525-7fa9235f0754)

i.e. `js` is grouped with all its aliases/extensions: `javascript, js, jsx, cjs, mjs, vue` and then `ts` / `tsx` is grouped with all its aliases/extensions: `typescript, ts, tsx, cts, mts, flow` and then `rust, rs` appear together, `yaml, yml` appear together etc.

possible values print out (note ordering):
```
error: invalid value 'javascrip' for '--language <LANGUAGE>'
  [possible values: javascript, js, jsx, cjs, mjs, vue, typescript, ts, tsx, cts, mts, flow, html, css, json, java, csharp, cs, python, py, ipynb, markdown, md, mdx, mdoc, markdowninline, go, rust, rs, ruby, rb, solidity, sol, hcl, tf, tfvars, yaml, yml, toml, sql, mysql, postgresql, php, phps, phar, phtml, pht, phponly]

  tip: a similar value exists: 'javascript'
```

which is more similar to how it prints out at the minute before my PR changes:

```
echo 'console.log(hello)' | cargo run -- apply '`hello` => `goodbye`' --stdin --lang=javascript
```
![image](https://github.com/user-attachments/assets/ac4fd253-aff3-4e3d-913c-55de2d0b6f0a)

possible values print out (note ordering):
```
error: invalid value 'javascript' for '--language <LANGUAGE>'
  [possible values: js, html, css, json, java, csharp, python, markdown, go, rust, ruby, solidity, hcl, yaml, sql, vue, toml, php, phponly]

  tip: a similar value exists: 'java'
```

if we go to the code snippet above, using only a `HashSet`, then it looks like this - which I think personally looks worse but if you prefer that I can certainly make the change:

```
echo 'console.log(hello)' | cargo run -- apply '`hello` => `goodbye`' --stdin --lang=javascrip
```

<img width="843" alt="image" src="https://github.com/user-attachments/assets/771c043d-6718-402c-ab98-fc15edac4272">

possible values print out (note ordering):
```
error: invalid value 'javascrip' for '--language <LANGUAGE>'
  [possible values: py, ipynb, phar, mts, css, mdx, tsx, mdoc, php, sol, phps, ruby, cts, jsx, md, rs, yaml, solidity, ts, json, yml, tfvars, flow, mysql, tf, vue, js, html, java, hcl, pht, markdowninline, go, postgresql, javascript, csharp, phponly, cs, rb, typescript, python, sql, mjs, markdown, rust, toml, cjs, phtml]

  tip: a similar value exists: 'javascript'
```

i.e. `py` and `ipynb` are together but `python` is "far away". Same for `rust` and `rs` as well as `yaml` and `yml` etc.

but if you are happy with this random ordering / shuffling I will make the change. Just let me know.

Reply:
Thanks, it makes sense why you did it this way.

As noted elsewhere, I actually don't want to include all the different flavors in the help. When providing help, providing many different variations of the same underlying value just pollutes the output.

If we switch it to just outputting the primary name for each language the hashset should be fine.

Reply:
The way that `PossibleValuesParser` works is different to this I believe ([docs.rs](https://docs.rs/clap/latest/clap/builder/struct.PossibleValuesParser.html)) it (currently) has to display all of them or none of them and any input needs to be one of these possible values but then you can map many-to-one like I had.

I would have to switch to a different approach to achieve what you're asking for and it would be a decent amount of work I think.

probably something like this: https://github.com/clap-rs/clap/issues/4416

i.e. `impl clap::ValueEnum for PatternLanguage` ([docs.rs](https://docs.rs/clap/latest/clap/builder/struct.EnumValueParser.html))

Unless you disagree, I'll try that out and see how hard it is.

Reply:
Yeah based on that discussion it looks like `ValueEnum` is what we are looking for. Specifically, allowing the aliases but not displaying them in help.

Reply:
wasn't too bad in the end - a bit more explicit and verbose but very clear what is happening: 20557f2